contracts/Removal.sol

```diff
@@ -80,37 +110,88 @@ contract Removal is
    * token id 1 URI points to vintage information (e.g., 2019) nori.com/api/removal/1 -> { amount: 10, supplier: 1, vintage: 2019, ... }
    * token id 2 URI points to vintage information (e.g., 2020) nori.com/api/removal/2 -> { amount: 50, supplier: 1, vintage: 2020, ... }
+   *
    * @param to The supplier address
    * @param amounts Each removal's tonnes of CO2 formatted as wei
    * @param ids The token ids to use for this batch of removals. The id itself encodes the supplier's ethereum address, a parcel identifier,
-   * the vintage, country code, state code, methodology identifer, and methodology version.
-   * @param data Encodes the market contract address and a unique identifier for the parcel from whence these removals came.
+   * the vintage, country code, state code, methodology identifer, methodology version, and id format.
+   * @param data Encodes the project id and schedule start time for this batch of removals, the market contract
+   * address and a boolean that indicates whether to list these removals for sale now.
    */
   function mintBatch(
     address to,
     uint256[] memory amounts,
     uint256[] memory ids,
     bytes memory data
   ) public override {
+    if (!(amounts.length == ids.length)) {
+      revert ArrayLengthMismatch({array1Name: "amounts", array2Name: "ids"});
+    }
+    // todo should we check that the removal id-encoded supplier address for each id is the same as `to` ?
+    // todo should we validate the removal ids any further? what about other malformed fields?
+    //    note that there is validation performed when creating ids in the first place, but room for errors/changes between id creation and then submission to this function
+    // todo do we add any validation to enforce that all removals in batch belong to the same project id?
     BatchMintRemovalsData memory decodedData = abi.decode(
       data,
       (BatchMintRemovalsData)
     );
+    if (decodedData.projectId == 0) {
+      revert InvalidProjectId({projectId: decodedData.projectId});
+    }
+    if (
+      decodedData.scheduleStartTime < UNIX_EPOCH_2010 ||
+      decodedData.scheduleStartTime > UNIX_EPOCH_2050
+    ) {
+      revert InvalidScheduleStartTime({
+        startTime: decodedData.scheduleStartTime
+      });
+    }
     for (uint256 i = 0; i < ids.length; i++) {
       if (_tokenIdExists[ids[i]]) {
-        revert TokenIdExists(); // todo can the duplicate token id be reported here?
+        // TODO (Gas Optimization): Is there a way to check once for all ids if they exist? e.g. via a merkle proof of some kind
+        revert TokenIdExists({tokenId: ids[i]});
       }
+      _removalIdToProjectId[ids[i]] = decodedData.projectId;
       _tokenIdExists[ids[i]] = true;
-      indexToTokenId[_tokenIdCounter] = ids[i];
-      _tokenIdCounter += 1;
+      indexToTokenId[tokenIdCounter] = ids[i];
+      tokenIdCounter += 1;
```

Comment:
We can save ~217 gas per iteration by accessing `ids[i]` once at the top of the for loop

```solidity
      uint256 id = ids[i]; // <--- access ids[i] once
      if (_tokenIdExists[id]) {
        revert TokenIdExists({tokenId: id});
      }
      _removalIdToProjectId[id] = decodedData.projectId;
      _tokenIdExists[id] = true;
      indexToTokenId[tokenIdCounter] = id; // todo use memory and assign once
      tokenIdCounter += 1;
```

Reply:
Save an additional 107 (1 removal) - 32,000 (in the case of 100 removals) gas by only referencing `tokenIdCounter` from before and after the loop
 ```solidity
    uint256 newTokenIdCounter = tokenIdCounter; // <-- hold it in memory instead of loading from storage each iteration
    for (uint256 i = 0; i < ids.length; i++) {
      uint256 id = ids[i];
      if (_tokenIdExists[id]) {
        revert TokenIdExists({tokenId: id});
      }
      _removalIdToProjectId[id] = decodedData.projectId;
      _tokenIdExists[id] = true;
      indexToTokenId[newTokenIdCounter] = id;
      newTokenIdCounter += 1;
    }
    tokenIdCounter = newTokenIdCounter;
 ```

Reply:
wow! I'm surprised the compiler can't optimize this away but I guess it makes sense that it doesn't know whether or not the storage slot *could* be accessed by another function call or something and therefore would be important to set with each iteration as opposed to once at the end.  nice catch, will update